### PR TITLE
Remove more characteristic 2 complications

### PR DESCRIPTION
**Pull request by botovq**
There's fix for a use after free in here, the rest is straightforward.

## Checklist
Thank you for your contribution. Please answer the following:

- [x] I acknowledge that I am authorized to submit this code under
the terms of the [Apache License](https://www.apache.org/licenses/LICENSE-2.0)
- [ ] All new public APIs are documented
- [ ] All new public functions have tests
- [ ] All new public functions have fuzzing tests

### CODE REVIEW

File: crypto/ec/ec_asn1.c [hunk 1]
```diff
@@ -204,30 +206,28 @@ static int ec_asn1_group2fieldid(const EC_GROUP *group, X9_62_FIELDID *field)
...
-            ERR_raise(ERR_LIB_EC, ERR_R_ASN1_LIB);
-            goto err;
-        }
-    } else if (nid == NID_X9_62_characteristic_two_field) {
+    if (nid == NID_X9_62_characteristic_two_field) {
         ERR_raise(ERR_LIB_EC, EC_R_GF2M_NOT_SUPPORTED);
         goto err;
-    }
-    else {
+    } else if (nid != NID_X9_62_prime_field) {
```
Comment (richsalz) [inline crypto/ec/ec_asn1.c:212]: Since it's now always `goto err` maybe turn the `else` into a nerwline?

File: crypto/ec/ec_backend.c [hunk 2]
```diff
@@ -158,21 +158,17 @@ static int ec_group_explicit_todata(const EC_GROUP *group, OSSL_PARAM_BLD *tmpl,
...
-
-    if (fid == NID_X9_62_prime_field) {
-        field_type = SN_X9_62_prime_field;
-    } else if (fid == NID_X9_62_characteristic_two_field) {
+    nid = EC_GROUP_get_field_type(group);
+    if (nid == NID_X9_62_characteristic_two_field) {
         ERR_raise(ERR_LIB_EC, EC_R_GF2M_NOT_SUPPORTED);
         goto err;
-    } else {
+    } else if (nid != NID_X9_62_prime_field) {
```
Comment (richsalz) [inline crypto/ec/ec_backend.c:171]: same, remove the else, start a new lne.

File: crypto/ec/ec_lib.c [hunk 3]
```diff
@@ -1501,13 +1485,11 @@ EC_GROUP *EC_GROUP_new_from_params(const OSSL_PARAM params[],
...
-    } else if (OPENSSL_strcasecmp(ptmp->data,
-                                  SN_X9_62_characteristic_two_field) == 0) {
-        is_prime_field = 0;
-    } else {
-        /* Invalid field */
+    if (OPENSSL_strcasecmp(ptmp->data,
+                           SN_X9_62_characteristic_two_field) == 0) {
+        ERR_raise(ERR_LIB_EC, EC_R_GF2M_NOT_SUPPORTED);
+        goto err;
+    } else if (OPENSSL_strcasecmp(ptmp->data, SN_X9_62_prime_field) != 0) {
```
Comment (richsalz) [inline crypto/ec/ec_lib.c:1492]: and here.

**Review by richsalz — CHANGES_REQUESTED** [review]
three nits.  looks good.  where was the use-after-free?

**Comment by botovq** [comment]
> where was the use-after-free?

It's right at the top, fixed in https://github.com/quictls/quictls/pull/177/commits/7f8ac72e39b1afc5c1f68740e7a0cb1c777e0433:
https://github.com/quictls/quictls/blob/4dbf2200e0ba7dd4301ddf1b16ec7260390d3ef2/crypto/ec/ec_asn1.c#L197-L204
If the field type is set to a nid without corresponding obj (since all methods are built-in and have a field type of `NID_X9_62_prime_field`, this isn't really possible), the goto err leaves a freed pointer in `ret->p.other`, making either the caller perform a double free in the cleanup path after
https://github.com/quictls/quictls/blob/4dbf2200e0ba7dd4301ddf1b16ec7260390d3ef2/crypto/ec/ec_asn1.c#L339-L341
or the caller of `EC_GROUP_get_ecparameters()` will receive a freed pointer in `params->p.other` and will then likely perform a double free when freeing `params` in its cleanup path.

**Review by richsalz — APPROVED** [review]
Looks like you made the `goto/else if` changes, thanks.

**Comment by botovq** [comment]
Yes, forgot to mention that. Thanks, merged.

**Pull request merged**